File: algorithm/127.word-ladder.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        count = 0
        word_set = set(wordList)

        if endWord not in word_set:
            return 0

        queue = [beginWord]

        while(len(queue) > 0):
            
            size = len(queue)
            count += 1

            for _i in range(size):
                node  = queue.pop(0)
            
                if node == endWord:
                    return count

                neighbors = self.get_neighbors(node, word_set)
                for neighbor in neighbors:
                    queue.append(neighbor)

        return 0

    # using 26 letters is better than check every nodes in wordList
    # l = length of word
    # 26 letters = O(l * 25 * l)
    # each word has l * 25 variation, and to check or change each attempt using l
    #
    # check every nodes in wordlist = O(l * N * N)
    # to check neighbors of every word in the list needs l * N, and there are N such words
    def get_neighbors(self, word, word_set):
        results = []
        for i in range(len(word)):
            for letter in 'abcdefghijklmnopqrstuvwxyz':
                if letter == word[i]:
                    continue

                neighbor = word[:i] + letter + word[(i + 1):]
                if neighbor not in word_set:
                    continue
                    
                results.append(neighbor)
                word_set.remove(neighbor) #this is important

        return results

File: leetcode/127.word-ladder.cpp
class Solution {
private:
    vector<string> getNeighbors(string& target, unordered_set<string>& words, unordered_set<string>& visited)
    {
        vector<string> res;
        string alphabet = "abcdefghijklmnopqrstuvwxyz";

        for (auto i = 0; i < target.size(); ++i)
        {
            for (auto j = 0; j < alphabet.size(); ++j)
            {
                string candidate = target;
                candidate[i] = alphabet[j];

                if (words.count(candidate) == 0)
                {
                    continue;
                }

                if (candidate == target)
                {
                    continue;
                }

                if (visited.count(candidate) > 0)
                {
                    continue;
                }

                res.push_back(candidate);
            }
        }

        return res;
    }
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> words;

        for (auto i = 0; i < wordList.size(); ++i)
        {
            words.insert(wordList[i]);
        }

        if (words.size() == 0 || words.count(endWord) == 0)
        {
            return 0;
        }

        queue<string> q;
        unordered_set<string> visited;

        q.push(beginWord);
        visited.insert(beginWord);
        int steps = 0;
        
        while(!q.empty())
        {
            steps++;
            int size = q.size();
            
            for (auto i = 0; i < size; ++i)
            {
                string tmp = q.front();
                q.pop();

                if (tmp == endWord)
                {
                    return steps;
                }

                for (auto neighbor: getNeighbors(tmp, words, visited))
                {
                    visited.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }

        return 0;
    }
};
